libs/client-shared/src/lib/utils/current-lang.ts

```diff
@@ -14,3 +18,28 @@ export function currentLangFactory(): Observable<string> {
     filter(isNotNil)
   );
 }
+
+export function getCurrentLang(): Observable<{ lang: string; path: any; queryParams: any }> {
+  const router = inject(Router);
+  return router.events.pipe(
+    filter((e): e is NavigationEnd => e instanceof NavigationEnd),
+    map((e) => e.urlAfterRedirects),
+    startWith(router.url),
+    map(
+      flow(
+        (url) => O.of(url.match('^/(\\w\\w)(.*)$')),
+        O.filter(isTruthy),
+        O.bindTo('match'),
+        O.bind('lang', ({ match }) => pipe(Lang.decode(match[1]), O.fromEither)),
+        O.bind('parsed', ({ match }) => O.of(queryString.parseUrl(match[2]))),
+        O.map(({ parsed, lang }) => ({
+          lang,
+          path: parsed.url,
+          queryParams: parsed.query,
+        }))
+      )
+    ),
+    map((it) => O.toNullable(it)),
+    filter(isNotNull)
+  );
+}
```

Comment:
I like the idea of this function, but I have some issues with it:

1. For every call to `getCurrentLang()`, we create a fully independent observable that listens to router events and processes them. Ideally, the observable is created only once, and then shared amongst all who need it. To do this, either place the observable in a service or store it in an [Injection Token](https://v17.angular.io/api/core/InjectionToken), and pipe its result into a [shareReplay](https://www.learnrxjs.io/learn-rxjs/operators/multicasting/sharereplay).
2. `path` can be typed with `string`, `queryParams` with `ParsedQuery<string>`.
3. The function's name reads as if the observable just returns the `lang` - I would rename it to something like `getParsedUrl`.

Reply:
The initial idea of this function was to reduce code duplication as the same code was used in the` language-selector` component and the `discalimer-dialog`. I now realised that there already is an INjectionToken for the current lang, which we can use in the dialog. This removes the need to have the `getCurrentLang()` (newly `getParsedUrl()`) method stored in a separate file and cna just be moved back to the `language-selector`.